### PR TITLE
Added frames processing to AnalysisBase

### LINKED ISSUE #1985: Allow analysis classes to run on arbitrary selections of frames
It would be convenient to be able to run some analyses on an arbitrary set of frames. I recently encounter a case where I wanted to build an RSMD matrix with `mda.analysis.diffusionmap.DistanceMatrix` on a selection of frame I obtained from clustering. For now, the only ways I could find to do that was to either write a new trajectory (on file or in memory), or to hack a `SelectionReader`. Adding a `frames` argument to run alonside `start`, `stop`, and `step` (see #1979) would allow to pass an array of frame indices, or a mask, or an already indexed trajectory using #1934. The interface would be the same as for `ag.write`.

```python
# API as envisioned

u = mda.Universe(TPR, XTC)
ag = u.select_atoms(...)

labels = my_clustering_method(ag)  # returns np.array( [1, 1, 0, 0, 2, ...])
frames_group0 = u.trajectory[labels == 0]

mat0 = MySuperAnalysis(ag).run(frames=frames_group0)
# or
mat0 = MySuperAnalysis(ag).run(frames=(labels == 0))

# Should prevent non-sensical calls such as mixing `frames` with `start`, `stop`, `step`.
mat0 = MySuperAnalysis(ag).run(frames=frames_group0, step=3)  # ValueError
```

Some analyses rely on the time step to be constant, so the use of the `frames` must be limited for these analyses. An slices trajectory can be used, but not an indexed one. In these case, a `NotImplementedError` should be raised. This could be achieved by using a class attribute to switch off arbitrary frame selections:

```python
class MyConstentDTAnalysis(AnalysisBase):
    needs_constant_dt = True

   ...
```

### PR DESCRIPTION
**Pull request by rsexton2**
Partially addresses #1985

Changes made in this Pull Request:
 - Make analysis base select arbitrary frames given by an array of integers (i.e., an array of frame indices) or booleans.
 
TODO (possibly), see #1985 
- accept FrameIterators

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]: I'd make this a bit more explicit "Added ability to iterate through trajectory by passing a list of frame indices".

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]: nitpicking, but I'd place frames before verbose here, mostly a visual thing but it looks nicer to have start/stop/step/frame bunched up together

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py:348]: Appropriate tests will be necessary here.

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py:357]: I _think_ using a frameiterator changes the behaviour of the iteration.

So if I do:

```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import TPR, XTC

u = mda.Universe(TPR, XTC)

iterator = u.trajectory[[0,1,2]]

for ts in iterator:
    pass

print(u.trajectory.ts)
```
Ouput: < Timestep 2 with unit cell dimensions [79.98359 79.98359 79.98358 60.      60.      90.     ] >

Whilst:

```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import TPR, XTC

u = mda.Universe(TPR, XTC)

for ts in u.trajectory[:3]
    pass

print(u.trajectory.ts)
```
Ouput: < Timestep 0 with unit cell dimensions [80.017006 80.017006 80.017006 60.       60.       90.      ] >

This is a general thing with the way we iterate through a trajectory based on an index selection (not limited to this code here), and we should probably fix this? But I'd say it's probably a bad idea for the two approaches differ like this for AnalysisBase.

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py:362]: I can see this becoming confusing real quickly, we're offering `frames` as an input argument to run and then offering a different thing contained in the `frames` user-facing variable. Thoughts?

Maybe we should call the input argument `indices` instead?

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py:396]: We'll need some documentation / examples for this.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of initial comments.

Main thing here is that I'm a little bit apprehensive about the differing behaviour about which frame you end up with at the end of for the frame iteration depending on if you use start/stop/step or a list of indices.

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/base.py:357]: I've opened a separate issue, I'm not actually sure if it's intended behaviour.

File: package/MDAnalysis/analysis/base.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/base.py:362]: I see your point. Although in principle `self.frames` should be equivalent to what the user inputs in `frames` in the sense that running a second time `a.run(frames=a.frames)` should give the identical answer (assuming that `ts.frame` is the true frame index), which I like. If the input `frames` was a boolean array or a `slice()` then `self.frames` will be an int array of frame indices but _should_ be functionally equivalent.

I am not too fond of calling the input `frames` kwarg `indices` because (I think) we typically use it for indexing atoms. (?)

File: package/MDAnalysis/analysis/base.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/base.py:357]: Thanks for opening  #3416.

As a band-aid solution we could add a `self._trajectory.rewind()` after the `for i, ts in enumerate(self._sliced_trajectory)` loop. This would ensure consistent behavior of all AnalysisBase classes. It would just need a comment that references issue #3416 so that we can remove it again if that gets fixed/changed.

File: package/MDAnalysis/analysis/base.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/base.py]: (I think that was me suggesting to keep the diff minimal... but you're absolutely right that frames should be with the other frame selections)

**Comment by orbeckst** [comment]
@rsexton2 please ping me when you're ready (remove draft status) for review and I am happy to have a look; while you're hacking away at tests, please already address @IAlibay's comments, which are all very good & relevant. Also have a look at what PEP8speaks complains about, please.

**Comment by rsexton2** [comment]
@orbeckst I added tests but still need to test slicing by a boolean array.
I think I addressed most of @IAlibay comments in the changes and fixed pep8 things

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: I changed the description to be more explicit, thank you.

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: I switched these around

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py:348]: I added tests to ensure frames isn't used with start, stop or step

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/base.py]: I would preferentially add the more explicit versionchanged text here rather than under `_setup_frames` since `run` is exposed in the docs but `_setup_frames` isn't.

Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: I added the versionchanged text to run

**Comment by orbeckst** [comment]
For some reason the CI was waiting for approval — sorry @rsexton2 , I didn't notice.

Given that you have added tests, do you want to set the status to "ready to review", i.e., remove the draft status?

File: package/MDAnalysis/analysis/base.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/base.py:362]: @IAlibay can you be convinced to leave the input `frames` as originally proposed or do you have strong opinions on the matter?

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_base.py:205]: Also add `[0, 3, 4, 3, 5]` with a repeated frame.

Otherwise I like that you're checking `an.frames` !

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_base.py]: What happens when you have `start=0, stop=0, step=0` ?

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_base.py]: remove the frames part from the parametrization. You never check against it, so this is just noise and confuses.

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_base.py]: Remove the reference value that is never tested
```suggestion
def test_frame_fail(u, run_kwargs):
```

**Review by orbeckst — CHANGES_REQUESTED** [review]
This looks already very good and your test coverage is great!

I only found minor things — I'll approve once you addressed my comments.

The only major remaining issue is some example docs as requested by @IAlibay and I will let him review that aspect (i.e., all reviewers must approve).

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (rsexton2) [reply testsuite/MDAnalysisTests/analysis/test_base.py:205]: Is the proper behavior that it will run frame 3 twice?

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (rsexton2) [reply testsuite/MDAnalysisTests/analysis/test_base.py]: done

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (rsexton2) [reply testsuite/MDAnalysisTests/analysis/test_base.py]: done

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (rsexton2) [reply testsuite/MDAnalysisTests/analysis/test_base.py]: This slipped through.
I changed _setup_frames to fail if start/stop/step aren't all None

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/analysis/test_base.py:205]: yes

**Review by orbeckst — APPROVED** [review]

File: .github/pr_path_labeler.yml
Comment (IAlibay) [inline .github/pr_path_labeler.yml]: Strange, why are these files here? Did the merge not go right?

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Can we use something similar to the versionchanged entry here? Whilst I'm sure it's not something most people read, it is where we direct folks to read about new changes, so being a bit more descriptive doesn't hurt.

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/base.py:357]: Sure I'm fine with this as a temporary measure. We should implement the fix in this PR so that we don't forget.

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/base.py:362]: It's a little bit messy but I'm not too fussed.

I'm not sure what (if anything) would need adding, but do we want to alter the attributes description for frame to make this clearer?

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_base.py:199]: It'd be too much to ask to add a test for every analysis method, but could we maybe throw in an extra test for RMSD that shows that passing frames works with that as expected?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Sorry only had time for a quick glance at it. Do ping me when you have the documentation written.

File: .github/pr_path_labeler.yml
Comment (orbeckst) [reply .github/pr_path_labeler.yml]: I'm not sure why they showed up. _Should_ be harmless when squash-merged, right?

**Comment by orbeckst** [comment]
@IAlibay I assigned the PR to you as you're the last one with outstanding change requests.

I don't know why some of the commits that are already in master show up here, perhaps they came from pushing the "update with develop" button (or whatever it's called when GitHub tells you a branch is out of date). I _think_ it will be fine when squash-merging but if there are any concerns then possibly rebasing this branch against and updated develop??

File: .github/pr_path_labeler.yml
Comment (rsexton2) [reply .github/pr_path_labeler.yml]: I rebased against an updated develop and squashed everything into 1 commit

File: package/CHANGELOG
Comment (rsexton2) [reply package/CHANGELOG]: Added more explicit description

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py:357]: I added the line to rewind with a comment referencing the issue

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py:396]: Added an example here with two uses of 'frames'

File: testsuite/MDAnalysisTests/analysis/test_base.py
Comment (rsexton2) [reply testsuite/MDAnalysisTests/analysis/test_base.py:199]: I added a test to test_rmsd.py

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py:362]: @IAlibay How should I alter the attribute description for `frames`? 
It is currently "array of Timestep frame indices. Only exists after calling AnalysisBase.run"
Something like "array of Timestep frame indices. Only exists after calling AnalysisBase.run. If the 'frames' 
option is used in .run(), the input for 'frames' will be returned"

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]: Thanks, could you just add a quick test to check that this is working as intended for now? (nothing fancy, just a quick check that the trajectory rewinds back to 0 will do)

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]: Should this be indented by two the right?

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]: Could you also throw in an example using start/stop/step? That way we have every use case here?

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/base.py:362]: With hindsight, let's leave the attribute as it is. I'm not really sure how to amend (or if it's at all possible) it without making it more complex.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Thanks, just a couple of hopefully final changes.

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: done

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: You're right, I fixed it

File: package/MDAnalysis/analysis/base.py
Comment (rsexton2) [reply package/MDAnalysis/analysis/base.py]: done

**Review by IAlibay — APPROVED** [review]
lgtm assuming CI returns green

**Comment by rsexton2** [comment]
git bisect says commit c35458b2b004a34a595c1c881be826cb9e0df460 is the one that makes the test testsuite/MDAnalysisTests/analysis/test_rdf.py::test_double_run FAILED

Locally testsuite/MDAnalysisTests/analysis/test_rdf.py::test_count FAILED
but not in the CI

```
c35458b2b004a34a595c1c881be826cb9e0df460 is the first bad commit
commit c35458b2b004a34a595c1c881be826cb9e0df460
Author: Ricky Sexton <rsexton2@asu.edu>
Date:   Fri Sep 10 15:49:19 2021 -0700

    Added rmsd test and 'frames' examples in docs

 package/CHANGELOG                              |  3 ++-
 package/MDAnalysis/analysis/base.py            | 18 ++++++++++++++++++
 testsuite/MDAnalysisTests/analysis/test_rms.py |  7 +++++++
 3 files changed, 27 insertions(+), 1 deletion(-)
```

The only thing that is added here is the rewind.
I'm not sure how this is affecting InterRDF

**Comment by rsexton2** [comment]
@richardjgowers @VOD555 
Do you have any idea what is happening with InterRDF and trajectory.rewind() ?

**Comment by IAlibay** [comment]
Just to follow up from discord - the issue seems to be from SingleFrameBase implementing an `_iter_` that doesn't rewind, which means that FrameIteratorAll rewinds on multi-frame trajectories but not on single frame readers.

In this particularly case - prior to this PR the `sels` fixture in test_rdf changes the coordinates and then keeps them that way. Whilst now the rewind specifically resets the timestep coordinates, hence yielding the test failure.

**Comment by orbeckst** [comment]
What do we need to do so that this PR can move forward?

**Comment by orbeckst** [comment]
As outlined in https://github.com/MDAnalysis/mdanalysis/pull/3415#issuecomment-922109775, iteration for single frame "trajectories" behaves differently from the way our multi-frame iterators _ought to_ behave (see also #3416 ). 

Could someone with proper understanding of the problem (@richardjgowers or @IAlibay ?) please raise an issue with a conclusion of what we will do about it? (I don't think it belongs into #3416, does it?) Then we can fix that problem and hopefully this PR can then be merged without additional changes.

**Comment by IAlibay** [comment]
> What do we need to do so that this PR can move forward?

Assuming we want to change the behaviour of singleframe, a quick fix here would be to just change the failing test to not use a function scoped fixture and instead just do the frame manipulation twice. Either way we'll need to do that once we implement the singleframe change/fix.

I'll raise an issue.

**Comment by jbarnoud** [comment]
I only had a very brief look, but since this PR already slices the trajectory, what makes it challenging to accept an already sliced trajectory as well?

If this PR goes through without accepting a sliced trajectory, the description of the PR should be edited so it does not close the original issue.

**Comment by orbeckst** [comment]
In the original discussion following https://github.com/MDAnalysis/mdanalysis/issues/1985#issuecomment-913865103 we had decided to start out conservatively with only explicit start/stop/step. I agree that being able to hand over FrameIterators would make it really complete. 

It makes sense exploring expanding this PR once we get the tests to pass, which requires #3642 IIRC.

**Comment by orbeckst** [comment]
For right now I changed the issue description to reflect the code that is present.

**Comment by orbeckst** [comment]
PR #3642 needs to get merged before we can finalize this PR.

File: package/AUTHORS
Comment (IAlibay) [inline package/AUTHORS]:
```suggestion
```
We don't usually have an empty line between year

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]:
```suggestion
        .. versionchanged:: 2.2.0
```

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]:
```suggestion
```
_should_ be fixed now

**Review by IAlibay — COMMENTED** [review]
I think just these and this is ready to go (going to self-merge to speed things up).

**Comment by IAlibay** [comment]
Assuming CI returns green I'm happy for this to get merged - @orbeckst do you want to have a final look?

**Comment by orbeckst** [comment]
Please go ahead at your leisure — don’t have the bandwidth and don’t want to hold up things. Thank you for looking after this one and the related PR!!

**Review by lilyminium — APPROVED** [review]
LGTM, thanks @rsexton2, @IAlibay, @orbeckst.

File: package/MDAnalysis/analysis/base.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/base.py]:
```suggestion

```

**Comment by IAlibay** [comment]
@orbeckst is this good to re-review or are you still working on the docs?

**Comment by orbeckst** [comment]
Done with docs, good for most final of final reviews (I hope)—feel free to just edit as you see fit. 

The main thing was to move a specific example out of run() because it would be really confusing for all tools to reference RMSD in their inherited run().

**Comment by orbeckst** [comment]
Can we merge this one ... finally??

**Comment by IAlibay** [comment]
> Can we merge this one ... finally??

Sorry it's on my task list, I'll have a final look in the morning

File: package/MDAnalysis/analysis/base.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/base.py]:
```suggestion
:mod:`MDAnalysis.analysis` module. They aim to be easily useable and
```

**Review by IAlibay — APPROVED** [review]
lgtm just the one word change which I'll commit

**Comment by IAlibay** [comment]
Thanks @rsexton2 ! Sorry it took so long to merge, but it fixed some pretty big bugs in MDAnalysis along the way! 🎉

**Pull request merged**
